### PR TITLE
From `chronic-probe-insertion` branch - Catching up README and cleanup to the main branch

### PR DESCRIPTION
**Pull request by ttngu207**

### CODE REVIEW

File: element_array_ephys/readers/spikeglx.py [hunk 1]
```diff
@@ -172,22 +173,28 @@ def __init__(self, meta_filepath):
...
 
-        self.chanmap = self._parse_chanmap(self.meta['~snsChanMap']) if '~snsChanMap' in self.meta else None
-        self.shankmap = self._parse_shankmap(self.meta['~snsShankMap']) if '~snsShankMap' in self.meta else None
-        self.imroTbl = self._parse_imrotbl(self.meta['~imroTbl']) if '~imroTbl' in self.meta else None
+        self.chanmap = (self._parse_chanmap(self.meta['~snsChanMap'])
+                        if '~snsChanMap' in self.meta else None)
+        self.shankmap = (self._parse_shankmap(self.meta['~snsShankMap'])
+                         if '~snsShankMap' in self.meta else None)
+        self.imroTbl = (self._parse_imrotbl(self.meta['~imroTbl'])
+                        if '~imroTbl' in self.meta else None)
```
Comment (dimitri-yatsenko) [inline element_array_ephys/readers/spikeglx.py:193]:
```suggestion
        self.chanmap = self._parse_chanmap(self.meta.get('~snsChanMap', None)
        self.shankmap = self._parse_shankmap(self.meta.get('~snsShankMap', None)
        self.imroTbl = self._parse_imrotbl(self.meta.get('~imroTbl', None)
```

File: element_array_ephys/readers/spikeglx.py [hunk 2]
```diff
@@ -276,12 +283,15 @@ def _parse_imrotbl(raw):
...
     def get_recording_channels_indices(self, exclude_sync=False):
         """
-        The indices of recorded channels (in chanmap) with respect to the channels listed in the imro table
+        The indices of recorded channels (in chanmap)
+         with respect to the channels listed in the imro table
         """
         recorded_chns_ind = [int(v[0]) for k, v in self.chanmap.items()
-                             if k != 'shape' and (not k.startswith('SY') if exclude_sync else True)]
+                             if k != 'shape'
+                             and (not k.startswith('SY') if exclude_sync else True)]
```
Comment (dimitri-yatsenko) [inline element_array_ephys/readers/spikeglx.py:291]:
```suggestion
                             and not (exclude_sync and k.startswith('SY'))]
```

**Review by shenshan — APPROVED** [review]

File: element_array_ephys/ephys.py [hunk 3]
```diff
@@ -0,0 +1,746 @@
...
+                    'spike_sites': spike_sites[kilosort_dataset.data['spike_clusters'] == unit],
+                    'spike_depths': spike_depths[kilosort_dataset.data['spike_clusters'] == unit]})
+
+        self.insert1(key)
+        self.Unit.insert([{**key, **u} for u in units])
+
+
+@schema
+class WaveformSet(dj.Imported):
+    definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:540]:
```suggestion
    definition = """
    # A set of spike waveforms for units out of a given CuratedClustering
```

File: element_array_ephys/ephys.py [hunk 4]
```diff
@@ -0,0 +1,746 @@
...
+    -> CuratedClustering
+    """
+
+    class PeakWaveform(dj.Part):
```
Comment (shenshan) [inline element_array_ephys/ephys.py:544]: I hate to suggest name changes at this point, but when reading, it feels like `UnitPeakWaveform` is a better one.

File: element_array_ephys/ephys.py [hunk 5]
```diff
@@ -0,0 +1,746 @@
...
+        definition = """
+        -> master
+        -> CuratedClustering.Unit
+        ---
+        peak_electrode_waveform: longblob  # (uV) mean waveform for this unit's peak electrode
+        """
+
+    class Waveform(dj.Part):
```
Comment (shenshan) [inline element_array_ephys/ephys.py:552]: UnitWaveform?

File: element_array_ephys/ephys.py [hunk 6]
```diff
@@ -0,0 +1,746 @@
...
+
+
+@schema
+class WaveformSet(dj.Imported):
+    definition = """
+    -> CuratedClustering
+    """
+
+    class PeakWaveform(dj.Part):
+        definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:545]:
```suggestion
        definition = """
        # Mean waveform across spikes for a given unit at the electrode with the peak spike amplitude.
```

File: element_array_ephys/ephys.py [hunk 7]
```diff
@@ -0,0 +1,746 @@
...
+        -> master
+        -> CuratedClustering.Unit
+        ---
+        peak_electrode_waveform: longblob  # (uV) mean waveform for this unit's peak electrode
```
Comment (shenshan) [inline element_array_ephys/ephys.py:549]:
```suggestion
        peak_electrode_waveform: longblob  # (uV) mean waveform across spikes for a given unit at the electrode with the peak spike amplitude.
```

File: element_array_ephys/ephys.py [hunk 8]
```diff
@@ -0,0 +1,746 @@
...
+        """
+
+    class Waveform(dj.Part):
+        definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:553]:
```suggestion
        definition = """
        # Spike waveforms and their mean across spikes for the given unit
```

File: element_array_ephys/ephys.py [hunk 9]
```diff
@@ -0,0 +1,746 @@
...
+        curation_id = dj.U().aggr(self & key, n='ifnull(max(curation_id)+1,1)').fetch1('n')
+        self.insert1({**key, 'curation_id': curation_id,
+                      'curation_time': creation_time, 'curation_output_dir': output_dir,
+                      'quality_control': is_qc, 'manual_curation': is_curated,
+                      'curation_note': curation_note})
+
+
+@schema
+class CuratedClustering(dj.Imported):
+    definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:470]:
```suggestion
    definition = """
    # Clustering results of a manual curation.
```

File: element_array_ephys/ephys.py [hunk 10]
```diff
@@ -0,0 +1,746 @@
...
+    -> Curation    
+    """
+
+    class Unit(dj.Part):
+        definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:475]:
```suggestion
        definition = """  
        # Unit properties after a manual curation. 
```

File: element_array_ephys/ephys.py [hunk 11]
```diff
@@ -0,0 +1,746 @@
...
+                                      ' clustering analysis is not yet supported')
+        else:
+            raise ValueError(f'Unknown task mode: {task_mode}')
+
+        self.insert1({**key, 'clustering_time': creation_time})
+
+
+@schema
+class Curation(dj.Manual):
+    definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:435]:
```suggestion
    definition = """
    # Manual curation procedure
```

File: element_array_ephys/ephys.py [hunk 12]
```diff
@@ -0,0 +1,746 @@
...
+
+@schema
+class Clustering(dj.Imported):
+    """
+    A processing table to handle each ClusteringTask:
+    + If `task_mode == "trigger"`: trigger clustering analysis
+        according to the ClusteringParamSet (e.g. launch a kilosort job)
+    + If `task_mode == "load"`: verify output
+    """
+    definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:409]:
```suggestion
    definition = """
    # Clustering Procedure
```

File: element_array_ephys/ephys.py [hunk 13]
```diff
@@ -0,0 +1,746 @@
...
+        ('good', 'single unit'),
+        ('ok', 'probably a single unit, but could be contaminated'),
+        ('mua', 'multi-unit activity'),
+        ('noise', 'bad unit')
+    ]
+
+
+@schema
+class ClusteringTask(dj.Manual):
+    definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:392]:
```suggestion
    definition = """
    # Manual table for a clustering task ready to be run
```

File: element_array_ephys/ephys.py [hunk 14]
```diff
@@ -0,0 +1,746 @@
...
+    clustering_method_desc: varchar(1000)
+    """
+
+    contents = [('kilosort', 'kilosort clustering method'),
+                ('kilosort2', 'kilosort2 clustering method')]
+
+
+@schema
+class ClusteringParamSet(dj.Lookup):
+    definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:342]:
```suggestion
    definition = """
    # Parameter set to be used in a clustering procedure
```

File: element_array_ephys/ephys.py [hunk 15]
```diff
@@ -0,0 +1,746 @@
...
+        # single insert in loop to mitigate potential memory issue
+        for electrode_key, lfp_trace in zip(electrode_keys, lfp):
+            self.Electrode.insert1({**key, **electrode_key, 'lfp': lfp_trace})
+
+
+# ------------ Clustering --------------
+
+@schema
+class ClusteringMethod(dj.Lookup):
+    definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:330]:
```suggestion
    definition = """
    # Method for clustering
```

File: element_array_ephys/ephys.py [hunk 16]
```diff
@@ -0,0 +1,746 @@
...
+@schema
+class AcquisitionSoftware(dj.Lookup):
+    definition = """  # Name of software used for recording of neuropixels probes - SpikeGLX or Open Ephys
+    acq_software: varchar(24)    
+    """
+    contents = zip(['SpikeGLX', 'Open Ephys'])
+
+
+@schema
+class ProbeInsertion(dj.Manual):  # (acute)
```
Comment (shenshan) [inline element_array_ephys/ephys.py:101]:
```suggestion
class ProbeInsertion(dj.Manual): 
```

File: element_array_ephys/ephys.py [hunk 17]
```diff
@@ -0,0 +1,746 @@
...
+    definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:102]:
```suggestion
    definition = """
    # Probe insertion implanted chronically into an animal.
```

File: element_array_ephys/ephys.py [hunk 18]
```diff
@@ -0,0 +1,746 @@
...
+    insertion_number: tinyint unsigned
+    ---
+    -> probe.Probe
+    insertion_datetime=null: datetime
+    """
+
+
+@schema
+class InsertionLocation(dj.Manual):
+    definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:113]:
```suggestion
    definition = """
    # Brain Location of a given probe insertion.
```

File: element_array_ephys/ephys.py [hunk 19]
```diff
@@ -0,0 +1,746 @@
...
+    depth:       decimal(6, 2) # (um) manipulator depth relative to surface of the brain (0); more ventral is more negative
+    theta=null:  decimal(5, 2) # (deg) - elevation - rotation about the ml-axis [0, 180] - w.r.t the z+ axis
+    phi=null:    decimal(5, 2) # (deg) - azimuth - rotation about the dv-axis [0, 360] - w.r.t the x+ axis
+    beta=null:   decimal(5, 2) # (deg) rotation about the shank of the probe [-180, 180] - clockwise is increasing in degree - 0 is the probe-front facing anterior
+    """
+
+
+@schema
+class EphysRecording(dj.Imported):
+    definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:128]:
```suggestion
    definition = """
    # Ephys recording out of a probe insertion at a given session.
```

File: element_array_ephys/ephys.py [hunk 20]
```diff
@@ -0,0 +1,746 @@
...
+    -> Session
+    -> ProbeInsertion      
+    ---
+    -> probe.ElectrodeConfig
+    -> AcquisitionSoftware
+    sampling_rate: float # (Hz) 
+    """
+
+    class EphysFile(dj.Part):
+        definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:138]:
```suggestion
        definition = """
        # Paths of files of a given Ephys session.
```

File: element_array_ephys/ephys.py [hunk 21]
```diff
@@ -0,0 +1,746 @@
...
+                                   for fp in probe_data.recording_info['recording_files']])
+        else:
+            raise NotImplementedError(f'Processing ephys files from'
+                                      f' acquisition software of type {acq_software} is'
+                                      f' not yet implemented')
+
+
+@schema
+class LFP(dj.Imported):
+    definition = """
```
Comment (shenshan) [inline element_array_ephys/ephys.py:238]:
```suggestion
    definition = """
    # Local field potential of a given Ephys recording.
```

**Review by shenshan — APPROVED** [review]

**Comment by ttngu207** [comment]
Thanks for the review @shenshan, many of these changes are also applicable at the `main` branch. I will take your suggestions and implement for the `main` branch and pull the changes to this branch

File: README.md [hunk 22]
```diff
@@ -24,32 +27,56 @@ and also requires some notion of ***Location*** as a dependency for ***Insertion
...
-    to supply a method to retrieve the directory containing the kilosort results, 
-    where the method's input arguments are the primary attributes identifying one ***Clustering***
-+ ***Unit*** - Identified unit(s) from one ***Clustering***, with associated ***ClusterQualityLabel***
-    + ***UnitSpikeTimes*** - spike times per unit
-    + ***Waveform*** - mean waveform across spikes per unit per recording electrode
++ ***Curation*** - specify instance(s) of curations performed on the output of a given ***Clustering***
++ ***CuratedClustering*** - set of results from a particular round of clustering/curation
+    + ***CuratedClustering.Unit*** - Identified unit(s) from one ***Curation***, and the associated properties (e.g. cluster quality, spike times, spike depths, etc.)
+    + ***WaveformSet*** - A set of spike waveforms for units from a given CuratedClustering
+
```
Comment (shenshan) [inline README.md:46]: Do we need explanation of part tables here too?

File: element_array_ephys/ephys.py [hunk 23]
```diff
@@ -0,0 +1,761 @@
...
+class CuratedClustering(dj.Imported):
+    definition = """
+    # Clustering results of a curation.
+    -> Curation    
+    """
+
+    class Unit(dj.Part):
+        definition = """   
+        -> master
+        # Properties of a given unit from a round of clustering (and curation)
```
Comment (shenshan) [inline element_array_ephys/ephys.py:488]: Should this comment go up one line?

**Review by shenshan — APPROVED** [review]

File: element_array_ephys/ephys.py (hunk 23 above)
Comment (ttngu207) [reply element_array_ephys/ephys.py:488]: thanks for the catch, fixed

File: README.md (hunk 22 above)
Comment (ttngu207) [reply README.md:46]: Yes, agreed, put perhaps on the `main` branch, so that the same changes can be reflected here as well (maybe in the next PRs)

**Pull request merged**